File: Hashing/L5_Unordered_map_in_C++STL.cpp
#include<iostream>
#include<unordered_map>
using namespace std;

int main(){
    unordered_map<string,int> m;
    // insert elements in random order
    m["gfg"]=20;
    m["ide"]=30;
    m["course"]=12;
    m["geeks"]=35;
    // m.insert({"dsd",24});

    // Iterate method 1           
    for(auto x:m)
        cout<<x.first<<" "<<x.second<<endl;
    // iterate method 2
    cout<<"\nMethod 2: To iterate the loop\n";
    for(auto itr=m.begin(); itr!=m.end(); ++itr)
         cout<<itr->first<<" "<<itr->second<<endl;
    // search elerment using find() function
    if(m.find("ide")!=m.end())
        cout<<"Found\n";
    else
        cout<<"Not Found\n";
    // search elerment using count() function
    if(m.count("rfj")>0)
        cout<<"Found\n";
    else
        cout<<"Not Found\n";
    
    cout<<"size "<<m.size()<<endl;

    m.erase("ide");
    cout<<"size after erase "<<m.size()<<endl;

    m.erase(m.begin()); // erase first element
    cout<<"size after erase begining "<<m.size()<<endl;

    m.erase(m.begin(),m.end()); // erase everything from start to last( m.end()-1)
    cout<<"size after erase everything "<<m.size()<<endl;
    return 0;
}

File: code_java/self pased DSA/Hashing/L5_Unordered_map_in_C++STL.java
/*
Problem: Demonstrate basic operations of unordered_map (C++)/HashMap (Java) such as insertion, iteration, search, erase, and size.
Approach:
- Use Java's HashMap to perform similar operations as C++ unordered_map.
*/
import java.util.*;

public class L5_Unordered_map_in_Java {
    public static void main(String[] args) {
        Map<String, Integer> m = new HashMap<>();
        // Insert elements
        m.put("gfg", 20);
        m.put("ide", 30);
        m.put("course", 12);
        m.put("geeks", 35);
        // Iterate method 1
        for (Map.Entry<String, Integer> x : m.entrySet())
            System.out.println(x.getKey() + " " + x.getValue());
        // Iterate method 2
        System.out.println("\nMethod 2: To iterate the loop");
        for (String key : m.keySet())
            System.out.println(key + " " + m.get(key));
        // Search element using containsKey
        if (m.containsKey("ide"))
            System.out.println("Found");
        else
            System.out.println("Not Found");
        // Search element using get
        if (m.get("rfj") != null)
            System.out.println("Found");
        else
            System.out.println("Not Found");
        System.out.println("size " + m.size());
        m.remove("ide");
        System.out.println("size after erase " + m.size());
        // Remove first element (no direct equivalent, so remove by key)
        Iterator<String> it = m.keySet().iterator();
        if (it.hasNext()) {
            String firstKey = it.next();
            m.remove(firstKey);
        }
        System.out.println("size after erase beginning " + m.size());
        m.clear();
        System.out.println("size after erase everything " + m.size());
    }
} 